// @ngrx/router-store/migrations/14_0_0/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const ts = __importStar(require("typescript"));
const schematics_1 = require("@angular-devkit/schematics");
const schematics_core_1 = require("../../schematics-core");
const renames = {
    DefaultRouterStateSerializer: 'FullRouterStateSerializer',
};
function renameSerializers() {
    return (tree) => {
        (0, schematics_core_1.visitTSSourceFiles)(tree, (sourceFile) => {
            const routerStoreImports = sourceFile.statements
                .filter(ts.isImportDeclaration)
                .filter(({ moduleSpecifier }) => moduleSpecifier.getText(sourceFile).includes('@ngrx/router-store'));
            if (routerStoreImports.length === 0) {
                return;
            }
            const changes = [
                ...findSerializerImportDeclarations(sourceFile, routerStoreImports),
                ...findSerializerReplacements(sourceFile),
            ];
            (0, schematics_core_1.commitChanges)(tree, sourceFile.fileName, changes);
        });
    };
}
function findSerializerImportDeclarations(sourceFile, imports) {
    const changes = imports
        .map((p) => p?.importClause?.namedBindings?.elements)
        .reduce((imports, curr) => imports.concat(curr ?? []), [])
        .map((specifier) => {
        if (!ts.isImportSpecifier(specifier)) {
            return { hit: false };
        }
        const serializerImports = Object.keys(renames);
        if (serializerImports.includes(specifier.name.text)) {
            return { hit: true, specifier, text: specifier.name.text };
        }
        // if import is renamed
        if (specifier.propertyName &&
            serializerImports.includes(specifier.propertyName.text)) {
            return { hit: true, specifier, text: specifier.propertyName.text };
        }
        return { hit: false };
    })
        .filter(({ hit }) => hit)
        .map(({ specifier, text }) => !!specifier && !!text
        ? (0, schematics_core_1.createReplaceChange)(sourceFile, specifier, text, renames[text])
        : undefined)
        .filter((change) => !!change);
    return changes;
}
function findSerializerReplacements(sourceFile) {
    const renameKeys = Object.keys(renames);
    const changes = [];
    ts.forEachChild(sourceFile, (node) => find(node, changes));
    return changes;
    function find(node, changes) {
        let change = undefined;
        if (ts.isPropertyAssignment(node) &&
            renameKeys.includes(node.initializer.getText(sourceFile))) {
            change = {
                node: node.initializer,
                text: node.initializer.getText(sourceFile),
            };
        }
        if (ts.isPropertyAccessExpression(node) &&
            renameKeys.includes(node.expression.getText(sourceFile))) {
            change = {
                node: node.expression,
                text: node.expression.getText(sourceFile),
            };
        }
        if (ts.isVariableDeclaration(node) &&
            node.type &&
            renameKeys.includes(node.type.getText(sourceFile))) {
            change = {
                node: node.type,
                text: node.type.getText(sourceFile),
            };
        }
        if (change) {
            changes.push((0, schematics_core_1.createReplaceChange)(sourceFile, change.node, change.text, renames[change.text]));
        }
        ts.forEachChild(node, (childNode) => find(childNode, changes));
    }
}
function default_1() {
    return (0, schematics_1.chain)([renameSerializers()]);
}
