// @ngrx/router-store/migrations/15_2_0/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const ts = __importStar(require("typescript"));
const schematics_1 = require("@angular-devkit/schematics");
const schematics_core_1 = require("../../schematics-core");
const renames = {
    getSelectors: 'getRouterSelectors',
};
function renameSelector() {
    return (tree) => {
        (0, schematics_core_1.visitTSSourceFiles)(tree, (sourceFile) => {
            const routerStoreImports = sourceFile.statements
                .filter((p) => ts.isImportDeclaration(p))
                .filter(({ moduleSpecifier }) => moduleSpecifier.getText(sourceFile).includes('@ngrx/router-store'));
            const changes = [
                ...replaceNamedImports(routerStoreImports, sourceFile),
                ...replaceNamespaceImports(routerStoreImports, sourceFile),
            ];
            if (changes.length) {
                (0, schematics_core_1.commitChanges)(tree, sourceFile.fileName, changes);
            }
        });
    };
}
function replaceNamedImports(routerStoreImports, sourceFile) {
    const changes = [];
    const namedImports = routerStoreImports
        .flatMap((p) => !!p.importClause && ts.isImportClause(p.importClause)
        ? p.importClause.namedBindings
        : [])
        .flatMap((p) => (!!p && ts.isNamedImports(p) ? p.elements : []));
    for (const namedImport of namedImports) {
        tryToAddReplacement(namedImport.name, sourceFile, changes);
    }
    return changes;
}
function replaceNamespaceImports(routerStoreImports, sourceFile) {
    const changes = [];
    const namespaceImports = routerStoreImports
        .map((p) => !!p.importClause &&
        ts.isImportClause(p.importClause) &&
        !!p.importClause.namedBindings &&
        ts.isNamespaceImport(p.importClause.namedBindings)
        ? p.importClause.namedBindings.name.getText(sourceFile)
        : null)
        .filter((p) => !!p);
    if (namespaceImports.length === 0) {
        return changes;
    }
    for (const statement of sourceFile.statements) {
        statement.forEachChild((child) => {
            if (ts.isVariableDeclarationList(child)) {
                const [declaration] = child.declarations;
                if (ts.isVariableDeclaration(declaration) &&
                    declaration.initializer &&
                    ts.isCallExpression(declaration.initializer) &&
                    declaration.initializer.expression &&
                    ts.isPropertyAccessExpression(declaration.initializer.expression) &&
                    ts.isIdentifier(declaration.initializer.expression.expression) &&
                    ts.isIdentifier(declaration.initializer.expression.name)) {
                    if (namespaceImports.includes(declaration.initializer.expression.expression.getText(sourceFile))) {
                        tryToAddReplacement(declaration.initializer.expression.name, sourceFile, changes);
                    }
                }
            }
        });
    }
    return changes;
}
function tryToAddReplacement(oldName, sourceFile, changes) {
    const oldNameText = oldName.getText(sourceFile);
    const newName = renames[oldNameText];
    if (newName) {
        const change = (0, schematics_core_1.createReplaceChange)(sourceFile, oldName, oldNameText, newName);
        changes.push(change);
    }
}
function default_1() {
    return (0, schematics_1.chain)([renameSelector()]);
}
